File: app/src/main/java/com/majeur/psclient/widget/SideView.java
package com.majeur.psclient.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.util.ArrayMap;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.View;

import com.majeur.psclient.model.Colors;
import com.majeur.psclient.util.Utils;

import java.util.Map;

import androidx.annotation.Nullable;

public class SideView extends View {

    private Map<String, Integer> mCurrentSides;
    private int mGravity;
    private int mWidth;

    private Paint mPaint;
    private Rect mTempRect;
    private Canvas mMockCanvas;
    private Point mMeasurePoint;
    private int mShadowRadius;
    private int mRectRadius;
    private int mTextSize;
    private int mVerticalMargin;

    public SideView(Context context) {
        this(context, null);
    }

    public SideView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SideView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mCurrentSides = new ArrayMap<>();
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.NORMAL));
        mTempRect = new Rect();
        mMockCanvas = new Canvas();
        mMeasurePoint = new Point();

        mShadowRadius = Utils.dpToPx(2);
        mVerticalMargin = Utils.dpToPx(4);
        mRectRadius = Utils.dpToPx(2);
        mTextSize = Utils.dpToPx(9);
    }

    public void setGravity(int gravity) {
        mGravity = gravity;
    }

    public void sideStart(String rawSide) {
        Log.e(getClass().getSimpleName(), "side start: " + rawSide);
        String key = rawSideToKey(rawSide);
        Integer currentCount = mCurrentSides.get(key);
        if (currentCount == null) currentCount = 0;
        mCurrentSides.put(key, currentCount + 1);
        requestLayout();
        invalidate();
    }

    public void sideEnd(String rawSide) {
        Log.e(getClass().getSimpleName(), "side end: " + rawSide);
        String key = rawSideToKey(rawSide);
        mCurrentSides.remove(key);
        requestLayout();
        invalidate();
    }

    public void clearAllSides() {
        mCurrentSides.clear();
        requestLayout();
        invalidate();
    }

    private String rawSideToKey(String rawSide) {
        rawSide = rawSide.trim();
        int spaceIndex = rawSide.indexOf(' ');
        if (spaceIndex >= 0)
            return (rawSide.substring(0, 1) + rawSide.substring(spaceIndex+1, spaceIndex+2))
                    .toUpperCase();
        return rawSide.substring(0, 2).toUpperCase();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mMeasurePoint.set(0, 0);
        drawContent(mMockCanvas, mMeasurePoint);
        setMeasuredDimension(mMeasurePoint.x, mMeasurePoint.y);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        mWidth = w;
        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        drawContent(canvas, null);
    }

    private void drawContent(Canvas canvas, Point measurePoint) {
        boolean gravityEnd = (mGravity & Gravity.END) == Gravity.END;
        int yOffset = mShadowRadius;
        for (Map.Entry<String, Integer> entry : mCurrentSides.entrySet()) {
            mPaint.setTextSize(mTextSize);
            int count = entry.getValue();
            String countString = (count > 1 ? (gravityEnd ? "" : " ") + "x" + count + (gravityEnd ? " " : ""): "");
            String text = gravityEnd ? countString + entry.getKey() : entry.getKey() + countString;
            mPaint.getTextBounds(text, 0, text.length(), mTempRect);

            mPaint.setShadowLayer(mShadowRadius, 0, mShadowRadius / 3, Colors.BLACK);
            mPaint.setColor(Colors.sideColor(entry.getKey()));
            int w = mTempRect.width() + 2 * mRectRadius + mShadowRadius;
            int h = mTempRect.height() + 2 * mRectRadius;
            if (gravityEnd) {
                canvas.drawRoundRect(mWidth - w + mShadowRadius, yOffset, mWidth, yOffset + h, mRectRadius, mRectRadius, mPaint);
                mPaint.clearShadowLayer();
                canvas.drawRect(mWidth - mRectRadius, yOffset, mWidth, yOffset + h, mPaint);
                mPaint.setColor(Colors.WHITE);
                canvas.drawText(text, mWidth - w + mRectRadius + mShadowRadius, mRectRadius + yOffset + mTempRect.height(), mPaint);
            } else {
                canvas.drawRoundRect(0, yOffset, w - mShadowRadius, yOffset + h, mRectRadius, mRectRadius, mPaint);
                mPaint.clearShadowLayer();
                canvas.drawRect(0, yOffset, mRectRadius, yOffset + h, mPaint);
                mPaint.setColor(Colors.WHITE);
                canvas.drawText(text, mRectRadius, mRectRadius + yOffset + mTempRect.height(), mPaint);
            }

            yOffset += h + mVerticalMargin;
            if (measurePoint != null) {
                if (measurePoint.x < w) measurePoint.x = w;
                if (measurePoint.y < yOffset) measurePoint.y = yOffset;
            }
        }
    }
}

File: psclient/src/main/java/com/majeur/psclient/widget/SideView.kt
package com.majeur.psclient.widget

import android.content.Context
import android.graphics.*
import android.view.Gravity
import android.view.View
import androidx.collection.ArrayMap
import com.majeur.psclient.model.common.Colors
import com.majeur.psclient.model.common.Colors.sideColor
import com.majeur.psclient.util.dp
import com.majeur.psclient.util.sp
import com.majeur.psclient.util.xForLeft
import com.majeur.psclient.util.yForTop
import kotlin.math.roundToInt

class SideView(context: Context?) : View(context) {

    var gravity = Gravity.NO_GRAVITY
        set(value) {
            field = value
            setPadding(if (isGravityRight) shadowRadius else 0, (shadowRadius - shadowDy).roundToInt(),
                    if (!isGravityRight) shadowRadius else 0, (shadowRadius + shadowDy).roundToInt())
        }
    private val isGravityRight
        get() = Gravity.getAbsoluteGravity(gravity, layoutDirection) and Gravity.HORIZONTAL_GRAVITY_MASK == Gravity.RIGHT

    private val sides = ArrayMap<String, Int>()
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var widthSize = 0
    private val path = Path()
    private val drawingRect = Rect()
    private val textBounds = Rect()
    private val tempRect by lazy { Rect() }
    private val shadowRadius = dp(2f)
    private val shadowDy = shadowRadius / 4f
    private val verticalSpacing = dp(4f)
    private val tagRadius = dp(2f)
    private val tagTextSize = sp(9f)

    init {
        gravity = Gravity.LEFT
        paint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.NORMAL)
    }
    
    fun sideStart(rawSide: String) {
        val key = rawSideToKey(rawSide)
        var currentCount = sides[key]
        if (currentCount == null) currentCount = 0
        sides[key] = currentCount + 1
        requestLayout()
        invalidate()
    }

    fun sideEnd(rawSide: String) {
        val key = rawSideToKey(rawSide)
        sides.remove(key)
        requestLayout()
        invalidate()
    }

    fun clearAllSides() {
        sides.clear()
        requestLayout()
        invalidate()
    }

    private fun rawSideToKey(rawSide: String): String {
        var rawSide = rawSide
        rawSide = rawSide.trim { it <= ' ' }
        val spaceIndex = rawSide.indexOf(' ')
        return if (spaceIndex >= 0) (rawSide.substring(0, 1) + rawSide.substring(spaceIndex + 1, spaceIndex + 2))
                .toUpperCase() else rawSide.substring(0, 2).toUpperCase()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        drawingRect.setEmpty()
        drawContent(null, drawingRect)
        setMeasuredDimension(drawingRect.width() + paddingLeft + paddingRight,
                drawingRect.height() + paddingTop + paddingBottom)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        widthSize = w
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawingRect.setEmpty()
        drawContent(canvas, drawingRect)
    }

    private fun drawContent(canvas: Canvas?, drawingRect: Rect) {
        var top = paddingTop
        for ((key, count) in sides) {
            val countString = if (count > 1) (if (isGravityRight) "" else " ") + "x" + count + (if (isGravityRight) " " else "") else ""
            val text = if (isGravityRight) countString + key else key + countString
            paint.apply {
                color = sideColor(key)
                textSize = tagTextSize.toFloat()
                getTextBounds(text, 0, text.length, textBounds)
                setShadowLayer(shadowRadius.toFloat(), 0f, shadowDy, Colors.BLACK)
            }

            val rectWith = textBounds.width() + 2 * tagRadius
            val rectHeight = textBounds.height() + 2 * tagRadius

            drawSemiRoundRect(canvas, tempRect,
                    if (isGravityRight) widthSize - paddingRight - rectWith else paddingLeft,
                    top,
                    if (isGravityRight) widthSize - paddingRight else paddingLeft + rectWith,
                    top + rectHeight,
                    isGravityRight)
            drawingRect.union(tempRect)
            paint.apply {
                color = Colors.WHITE
                clearShadowLayer()
            }
            canvas?.drawText(text,
                    textBounds.xForLeft(
                            if (isGravityRight) widthSize - paddingRight - rectWith + tagRadius else paddingLeft + tagRadius
                    ).toFloat(), textBounds.yForTop(top + tagRadius).toFloat(), paint)
            top += rectHeight + verticalSpacing
        }
    }

    private fun drawSemiRoundRect(canvas: Canvas?, drawingRect: Rect, left: Int, top: Int, right: Int, bottom: Int, flatAtRight: Boolean) {
        path.reset()
        if (flatAtRight) path.apply {
            moveTo(right.toFloat(), top.toFloat())
            lineTo(left + tagRadius.toFloat(), top.toFloat())
            quadTo(left.toFloat(), top.toFloat(), left.toFloat(), top + tagRadius.toFloat())
            lineTo(left.toFloat(), bottom - tagRadius.toFloat())
            quadTo(left.toFloat(), bottom.toFloat(), left + tagRadius.toFloat(), bottom.toFloat())
            lineTo(right.toFloat(), bottom.toFloat())
            lineTo(right.toFloat(), top.toFloat())
        } else path.apply {
            moveTo(left.toFloat(), top.toFloat())
            lineTo(right - tagRadius.toFloat(), top.toFloat())
            quadTo(right.toFloat(), top.toFloat(), right.toFloat(), top + tagRadius.toFloat())
            lineTo(right.toFloat(), bottom - tagRadius.toFloat())
            quadTo(right.toFloat(), bottom.toFloat(), right - tagRadius.toFloat(), bottom.toFloat())
            lineTo(left.toFloat(), bottom.toFloat())
            lineTo(left.toFloat(), top.toFloat())
        }
        canvas?.drawPath(path, paint)
        drawingRect.set(left, top, right, bottom)
    }
}